packages/server/lib/utils/utils.ts

```diff
@@ -43,6 +43,15 @@ export function missesInterpolationParam(str: string, replacers: Record<string,
     return /\${([^{}]*)}/g.test(interpolatedStr);
 }
 
+/**
+ * A helper function to check if any string in an object misses interpolation params.
+ * For example:
+ * missesInterpolationParamInObject({ context: 'stores/${storeHash}', response_type: 'code' }, { storeHash: 'abc123' }) -> returns false
+ * missesInterpolationParamInObject({ context: 'stores/${storeHash}', response_type: 'code' }, {}) -> returns true
+ */
+export function missesInterpolationParamInObject(params: Record<string, string>, replacers: Record<string, any>) {
+    return Object.values(params).some((param) => missesInterpolationParam(param, replacers));
+}
```

Comment:
not for this PR but I would prefer to see a single unit-tested function `interpolate` that support nested content and return an error when missing param instead of having 4 or 5 flavors of the same thing

Reply:
Yes, I agree that we should refactor this to reduce redundancy 👍.